### PR TITLE
Feature/python model v1

### PR DESCRIPTION
**Pull request by ChenyuLInx**
This change currently includes table materialization.

Also super happy to hear any feedback and what you think we missed.
TODO:
- [x] pass in timeout, retry for [dataproc submit](https://github.com/dbt-labs/dbt-bigquery/pull/224#discussion_r925052294) leaving it it out and commented in #231 to revisit
- [x] [Add clear error message](https://github.com/dbt-labs/dbt-bigquery/pull/224#discussion_r926051917) when running python model but related library is not installed

### CODE REVIEW

File: tests/functional/adapter/test_python.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    dbt.config(
+        materialized='table',
+    )
+    data = [[1,2]] * 10
+    return spark.createDataFrame(data, schema=['test', 'test2'])
+"""
+models__simple_python_model_v2 = """
+import pandas
+
+def model(dbt):
```
Comment (ChenyuLInx) [inline tests/functional/adapter/test_python.py:21]: What the model code looks like

File: dbt/include/bigquery/macros/materializations/table.sql [hunk 2]
```diff
@@ -35,3 +38,29 @@
...
+-- TODO dataproc requires a temp bucket to perform BQ write
+-- this is hard coded to internal testing ATM. need to adjust to render
+-- or find another way around
+{% macro py_complete_script(compiled_code, target_relation) %}
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder.appName('smallTest').getOrCreate()
+
+spark.conf.set("viewsEnabled","true")
+spark.conf.set("temporaryGcsBucket","python-model-test")
```
Comment (ChenyuLInx) [inline dbt/include/bigquery/macros/materializations/table.sql:51]: TODO in the comment above

File: dbt/adapters/bigquery/impl.py [hunk 3]
```diff
@@ -21,11 +22,19 @@
 import google.cloud.exceptions
 import google.cloud.bigquery
 
+try:
+    # Library only needed for python models
+    from google.cloud import dataproc_v1
+    from google.cloud import storage
```
Comment (jtcohen6) [inline dbt/adapters/bigquery/impl.py:28]: - Let's add these as an `extras_require` — e.g. `pip install dbt-bigquery[dataproc]`
- Let's raise a clearer error below if a user wants to use Python models, and hasn't installed the extra

File: dbt/adapters/bigquery/impl.py [hunk 4]
```diff
@@ -803,3 +812,84 @@ def run_sql_for_tests(self, sql, fetch, conn=None):
...
+            "pyspark_job": {"main_python_file_uri": "gs://{}/{}".format(self.credential.gcs_bucket, filename)},
+        }
+        operation = job_client.submit_job_as_operation(
+            request={"project_id": self.credential.database, "region": self.credential.dataproc_region, "job": job}
+        )
+        response = operation.result()
+        # Dataproc job output is saved to the Cloud Storage bucket
+        # allocated to the job. Use regex to obtain the bucket and blob info.
+        matches = re.match("gs://(.*?)/(.*)", response.driver_output_resource_uri)
+        output = (
```
[the remark this answers is not in the document]
Comment (ChenyuLInx) [reply dbt/adapters/bigquery/impl.py:889]: Yes, I looked into it briefly, didn't found much things to surface up right now. We can take another look later on

File: dbt/adapters/bigquery/impl.py [hunk 5]
```diff
@@ -803,3 +812,84 @@ def run_sql_for_tests(self, sql, fetch, conn=None):
...
+            storage.Client()
+            .get_bucket(matches.group(1))
+            .blob(f"{matches.group(2)}.000000000")
+            .download_as_string()
+        )
```
[the remark this answers is not in the document]
Comment (jtcohen6) [reply dbt/adapters/bigquery/impl.py:894]: ope ok makes sense! :)

**Review by stu-k — APPROVED** [review]
Nothing stands out 👍

**Review by gshank — APPROVED** [review]
I don't see why the dataproc_region wasn't found. Once that is working, looks fine.

**Pull request merged**